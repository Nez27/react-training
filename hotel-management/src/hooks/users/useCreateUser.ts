import { useMutation, useQueryClient } from '@tanstack/react-query';
import toast from 'react-hot-toast';

// Services
import { createUser as createUserFn } from '@service/userServices';

// Constants
import { ADD_SUCCESS } from '@constant/messages';

// Hooks
import { useUserRoomAvailable } from '@hook/useUserRoomAvailable';

/**
 * Create user in database
 * @returns The status of creating users and createUser function
 */
const useCreateUser = () => {
  const queryClient = useQueryClient();
  const { dispatch } = useUserRoomAvailable();

  const { mutate: createUser, isPending: isCreating } = useMutation({
    mutationFn: createUserFn,
    onSuccess: (user) => {
      toast.success(ADD_SUCCESS);
      queryClient.invalidateQueries({ queryKey: ['users'] });

      // Add user to global statement
      dispatch!({
        type: 'addUser',
        payload: [{ id: user.id, name: user.name }],
      });
    },
    onError: (err) => toast.error(err.message),
  });

  return { isCreating, createUser };
};

export { useCreateUser };
