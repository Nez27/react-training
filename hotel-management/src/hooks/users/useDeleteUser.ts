import { useMutation, useQueryClient } from '@tanstack/react-query';
import toast from 'react-hot-toast';

// Services
import { deleteUser as deleteUserFn } from '@service/userServices';

// Constants
import * as messages from '@constant/messages';

/**
 * Delete the user from database
 * @returns The boolean of status deleting and deleteUser function
 */
const useDeleteUser = () => {
  const queryClient = useQueryClient();

  const {
    isPending: isDeleting,
    mutate: deleteUser,
    isSuccess
  } = useMutation({
    mutationFn: deleteUserFn,
    onSuccess: () => {
      toast.success(messages.DELETE_SUCCESS);
      queryClient.invalidateQueries({
        queryKey: ['users'],
      });
    },
    onError: (err) => toast.error(err.message),
  });

  return { isDeleting, deleteUser, isSuccess };
};

export { useDeleteUser };
