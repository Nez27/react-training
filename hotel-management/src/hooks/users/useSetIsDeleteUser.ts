import { useMutation, useQueryClient } from '@tanstack/react-query';
import toast from 'react-hot-toast';

// Services
import { setIsDeleteUser as setIsDeleteUserFn } from '@service/userServices';

// Constants
import * as messages from '@constant/messages';
import { useUserRoomAvailable } from '@hook/useUserRoomAvailable';

/**
 * Delete the user from database
 * @returns The boolean of status deleting and deleteUser function
 */
const useIsDeleteUser = () => {
  const queryClient = useQueryClient();
  const { dispatch } = useUserRoomAvailable();

  const {
    isPending: isDeleting,
    mutate: setIsDeleteUser,
    isSuccess,
  } = useMutation({
    mutationFn: setIsDeleteUserFn,
    onSuccess: (user) => {
      toast.success(messages.DELETE_SUCCESS);
      queryClient.invalidateQueries({
        queryKey: ['users'],
      });

      // Remove user from global state
      dispatch?.({ type: 'removeUser', payload: [{ id: user.id }] });
    },
    onError: (err) => toast.error(err.message),
  });

  return { isDeleting, setIsDeleteUser, isSuccess };
};

export { useIsDeleteUser };
