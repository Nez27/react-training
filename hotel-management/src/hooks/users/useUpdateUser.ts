import toast from 'react-hot-toast';
import { useMutation, useQueryClient } from '@tanstack/react-query';

// Services
import { updateUser as updateUserFn } from '@service/userServices';

// Messages
import { UPDATE_SUCCESS } from '@constant/messages';

// Hooks
import { useUserRoomAvailable } from '@hook/useUserRoomAvailable';

/**
 * Update user from database
 * @returns The status of updating user and updateUser function
 */
const useUpdateUser = () => {
  const queryClient = useQueryClient();
  const { dispatch } = useUserRoomAvailable();

  const { mutate: updateUser, isPending: isUpdating } = useMutation({
    mutationFn: updateUserFn,
    onSuccess: (user) => {
      toast.success(UPDATE_SUCCESS);
      queryClient.invalidateQueries({ queryKey: ['users'] });

      // Update name in global state
      dispatch!({
        type: 'updateUserName',
        payload: [{ id: user.id, name: user.name }],
      });
    },
    onError: (err) => toast.error(err.message),
  });

  return { isUpdating, updateUser };
};

export { useUpdateUser };
