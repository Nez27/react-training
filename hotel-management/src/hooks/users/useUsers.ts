import toast from 'react-hot-toast';
import { useQuery, useQueryClient } from '@tanstack/react-query';
import { useSearchParams } from 'react-router-dom';

// Services
import { getAllUsers } from '@src/services/userServices';

// Constants
import { DEFAULT_PAGE_SIZE } from '@src/constants/config';

/**
 * Fetch data of users from database
 * @returns The status of loading users and data of users
 */
const useUsers = () => {
  const [searchParams] = useSearchParams();
  const queryClient = useQueryClient();
  const sortByValue = searchParams.get('sortBy') || 'id';
  const orderByValue = searchParams.get('orderBy') || 'asc';
  const searchValue = searchParams.get('search') || '';
  const page = searchParams.get('page')
    ? Number(searchParams.get('page'))
    : 1;

  const {
    isLoading,
    data: { data: users, count } = {},
    error,
  } = useQuery({
    queryKey: ['users', sortByValue, orderByValue, searchValue, page],
    queryFn: () =>
      getAllUsers({
        sortBy: sortByValue,
        orderBy: orderByValue,
        phoneSearch: searchValue,
        page,
      }),
  });

  if (error) {
    toast.error(error.message);
  }

  // Pre-fetching
  const totalPage = Math.ceil(count! / DEFAULT_PAGE_SIZE);

  if (page < totalPage) {
    const nextPage = page + 1;

    queryClient.prefetchQuery({
      queryKey: ['users', sortByValue, orderByValue, searchValue, nextPage],
      queryFn: () =>
        getAllUsers({
          sortBy: sortByValue,
          orderBy: orderByValue,
          phoneSearch: searchValue,
          page: nextPage,
        }),
    });
  }

  if (page > 1) {
    const previousPage = page - 1;

    queryClient.prefetchQuery({
      queryKey: ['users', sortByValue, orderByValue, searchValue, previousPage],
      queryFn: () =>
        getAllUsers({
          sortBy: sortByValue,
          orderBy: orderByValue,
          phoneSearch: searchValue,
          page: previousPage,
        }),
    });
  }

  return {
    isLoading,
    users,
    count,
  };
};

export { useUsers };
