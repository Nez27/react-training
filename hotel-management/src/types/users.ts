interface IUser {
  id: number;
  name: string;
  phone: string;
}

export type { IUser };
