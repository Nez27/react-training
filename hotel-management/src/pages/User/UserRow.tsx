import { useCallback } from 'react';

// Components
import { RiEditBoxFill } from 'react-icons/ri';
import Modal from '@component/Modal';
import Table from '@component/Table';
import Menus from '@component/Menus';
import UserForm from './UserForm';
import ConfirmMessage from '@component/ConfirmMessage';
import { HiTrash } from 'react-icons/hi';

// Types
import { IUser } from '@type/user';

// Constants
import { FORM } from '@constant/commons';

// Hooks
import { useIsDeleteUser } from '@hook/users/useSetIsDeleteUser';

interface IUserRow {
  user: IUser;
}

const UserRow = ({ user }: IUserRow) => {
  const {
    id,
    name,
    phone,
    isBooked
  } = user;
  const { setIsDeleteUser } = useIsDeleteUser();

  const renderEditBtn = useCallback(
    (onCloseModal: () => void) => (
      <Menus.Button onClick={onCloseModal} icon={<RiEditBoxFill />}>
        Edit
      </Menus.Button>
    ),
    []
  );

  const renderDeleteBtn = useCallback(
    (onCloseModal: () => void) => (
      <Menus.Button icon={<HiTrash />} onClick={onCloseModal} disabled={isBooked}>
        Delete
      </Menus.Button>
    ),
    [isBooked]
  );
  return (
    <Table.Row>
      <div>{id}</div>
      <div>{name}</div>
      <div>{phone}</div>
      <div>{
        isBooked 
          ? 'Yes' 
          : 'No'
          }</div>
      <div>
        <Modal>
          <Menus.Menu>
            <Menus.Toggle id={id.toString()} />

            <Menus.List id={id.toString()}>
              <Modal.Open
                modalName={FORM.EDIT}
                renderChildren={renderEditBtn}
              />

              <Modal.Open
                modalName={FORM.DELETE}
                renderChildren={renderDeleteBtn}
              />
            </Menus.List>

            <Modal.Window name={FORM.EDIT} title="Edit user">
              <UserForm user={user} />
            </Modal.Window>

            <Modal.Window name={FORM.DELETE} title="Delete User">
              <ConfirmMessage
                message={`Are you sure to delete ${name}?`}
                onConfirm={() => setIsDeleteUser(id)}
              />
            </Modal.Window>
          </Menus.Menu>
        </Modal>
      </div>
    </Table.Row>
  );
};

export default UserRow;
