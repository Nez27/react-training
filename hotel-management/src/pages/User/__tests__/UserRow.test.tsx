import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import renderer from 'react-test-renderer';

// Types
import { IUser } from '@type/users';

// Components
import UserRow from '../UserRow';

describe('UserRow', () => {
  const tempUser: IUser = {
    id: 1,
    name: 'Temp Room',
    phone: '0324421232',
    isBooked: false,
  };
  const queryClient = new QueryClient();
  const wrapper = renderer.create(
    <QueryClientProvider client={queryClient}>
      <UserRow user={tempUser} />
    </QueryClientProvider>
  );

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
