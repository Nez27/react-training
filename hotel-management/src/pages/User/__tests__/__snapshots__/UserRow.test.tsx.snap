// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRow Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-jSFipO sc-iCoHzw hTfHNc eWZMfB"
      >
        <div>
          1
        </div>
        <div>
          Temp Room
        </div>
        <div>
          0324421232
        </div>
        <div>
          No
        </div>
        <div>
          <div
            class="sc-kEqXeH jLboIv"
          >
            <button
              aria-label="Menu item 1"
              class="sc-crzpnZ gdsCet"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M10.5 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zm0 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zm0 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0z"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jSFipO sc-iCoHzw hTfHNc eWZMfB"
    >
      <div>
        1
      </div>
      <div>
        Temp Room
      </div>
      <div>
        0324421232
      </div>
      <div>
        No
      </div>
      <div>
        <div
          class="sc-kEqXeH jLboIv"
        >
          <button
            aria-label="Menu item 1"
            class="sc-crzpnZ gdsCet"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M10.5 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zm0 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zm0 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
